initial draft of dump doc clarification (#32988)
dotnet/docs: docs/core/diagnostics/collect-dumps-crash.md

---
title: Collect Dumps on Crash - .NET
description: How to collect a dump on a crash
ms.date: 12/13/2022
---

# Collect dumps on crash

Configuring your application to collect a dump on crash is done by setting specific environment variables. This is helpful when you want to get an understanding of why a crash happened. For example, capturing a dump when an exception is thrown helps you identify an issue by examining the state of the app when it crashed.

The following table shows the environment variables you can configure for collecting dumps on a crash.

|Environment variable|Description|Default value|
|-------|---------|---|
|`COMPlus_DbgEnableMiniDump` or `DOTNET_DbgEnableMiniDump`|If set to 1, enable core dump generation.|0|

> |`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see the table below|2 (`MiniDumpWithPrivateReadWriteMemory`)|

Author: It turns into a table: 
<img width="818" alt="image" src="https://user-images.githubusercontent.com/45022607/215917461-a43e910e-0e9b-4419-acad-e4d407202455.png">

> |`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see the table below|2 (`MiniDumpWithPrivateReadWriteMemory`)|

Reviewer: Are you referring to the next row then?

> |`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see the table below|2 (`MiniDumpWithPrivateReadWriteMemory`)|

Author: Row 13 is the start of the table I am referring to.

> |`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see the table below|2 (`MiniDumpWithPrivateReadWriteMemory`)|

Reviewer: ```suggestion
|`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see [Types of mini dumps](#types-of-mini-dumps).|2 (`MiniDumpWithPrivateReadWriteMemory`)|
```

Reviewer: I'm also confused why 2 corresponds to `MiniDumpWithPrivateReadWriteMemory` here, but in the referenced table, it corresponds to `Heap`. Are there two different ways to write it?

Author: @noahfalk would you please lend some insight to this?

Reviewer 2: Those two names are refering to the same concept and I'd recommend we replace MiniDumpWithPrivateReadWriteMemory -> Heap for naming consistency.

I believe what happened is in the past the doc used those very long names everywhere which correspond to some constants defined on the win32 MiniDumpWriteDump API. However someone probably recognized there was little value in using complicated API naming details when explaining these concepts. When they refactored to simpler names this reference probably got overlooked.

> |`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see the table below|2 (`MiniDumpWithPrivateReadWriteMemory`)|

Author: It is referring to the minidump table below the table that the environment variables are in.

ms.date: 12/13/2022
---

# Collect dumps on crash

Configuring your application to collect a dump on crash is done by setting specific environment variables. This is helpful when you want to get an understanding of why a crash happened. For example, capturing a dump when an exception is thrown helps you identify an issue by examining the state of the app when it crashed.

The following table shows the environment variables you can configure for collecting dumps on a crash.

|Environment variable|Description|Default value|
|-------|---------|---|
|`COMPlus_DbgEnableMiniDump` or `DOTNET_DbgEnableMiniDump`|If set to 1, enable core dump generation.|0|
|`COMPlus_DbgMiniDumpType` or `DOTNET_DbgMiniDumpType`|Type of dump to be collected. For more information, see [Types of mini dumps](#types-of-mini-dumps).|2 (`Heap`)|
|`COMPlus_DbgMiniDumpName` or `DOTNET_DbgMiniDumpName`|Path to a file to write the dump to. Ensure that the user under which the dotnet process is running has write permissions to the specified directory.|`/tmp/coredump.<pid>`|
|`COMPlus_CreateDumpDiagnostics` or `DOTNET_CreateDumpDiagnostics`|If set to 1, enables diagnostic logging of dump process.|0|

> |`COMPlus_EnableCrashReport` or `DOTNET_EnableCrashReport`|(Requires .NET 6 or later, not supported on Windows) If set to 1, the runtime generates a JSON-formatted crash report that includes information about the threads and stack frames of the crashing application. The crash report name is the dump path or name with *.crashreport.json* appended.

Reviewer: ```suggestion
|`COMPlus_EnableCrashReport` or `DOTNET_EnableCrashReport`|(Requires .NET 6 or later; not supported on Windows.)<br/>If set to 1, the runtime generates a JSON-formatted crash report that includes information about the threads and stack frames of the crashing application. The crash report name is the dump path or name with *.crashreport.json* appended.
```

> |`COMPlus_CreateDumpVerboseDiagnostics` or `DOTNET_CreateDumpVerboseDiagnostics`|(Requires .NET 7 or later) If set to 1, enables verbose diagnostic logging of the dump process.|0|

Reviewer: ```suggestion
|`COMPlus_CreateDumpVerboseDiagnostics` or `DOTNET_CreateDumpVerboseDiagnostics`|(Requires .NET 7 or later.)<br/>If set to 1, enables verbose diagnostic logging of the dump process.|0|
```

> |`COMPlus_CreateDumpLogToFile` or `DOTNET_CreateDumpLogToFile`|(Requires .NET 7 or later) The path of the file to which the diagnostic messages should be written. If unset, the diagnostic messages are written to the console of the crashing application.|

Reviewer: ```suggestion
|`COMPlus_CreateDumpLogToFile` or `DOTNET_CreateDumpLogToFile`|(Requires .NET 7 or later.)<br/>The path of the file to which the diagnostic messages should be written. If unset, the diagnostic messages are written to the console of the crashing application.|
```

> [!NOTE]
> .NET 7 standardizes on the prefix `DOTNET_` instead of `COMPlus_` for these environment variables. However, the `COMPlus_` prefix will continue to work. If you're using a previous version of the .NET runtime, you should still use the `COMPlus_` prefix for environment variables.

## File path templates

Starting in .NET 5, `DOTNET_DbgMiniDumpName` can also include formatting template specifiers that will be filled in dynamically:

|Specifier|Value|
|---------|-----|
|%%|A single % character|
|%p|PID of dumped process|
|%e|The process executable filename|
|%h|Host name return by `gethostname()`|
|%t|Time of dump, expressed as seconds since the Epoch, 1970-01-01 00:00:00 +0000 (UTC)|

## Types of mini dumps

> The following table shows all the values you can use for `DOTNET_DbgMiniDumpType`. For example, setting `DOTNET_DbgMiniDumpType` to 1 means `MiniDumpNormal` type dump will be collected on a crash.

Reviewer: ```suggestion
The following table shows all the values you can use for `DOTNET_DbgMiniDumpType`. For example, setting `DOTNET_DbgMiniDumpType` to 1 means `Mini` type dump will be collected on a crash.
```
